src/audio/tdfb/tdfb.c

```diff
@@ -147,7 +176,103 @@ static int tdfb_init_coef(struct tdfb_comp_data *cd, int source_nch,
 		return -EINVAL;
 	}
 
-	coefp = ASSUME_ALIGNED(&config->data[0], 2);
+	if (config->num_angles > SOF_TDFB_MAX_ANGLES) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid num_angles %d",
+			    config->num_angles);
+		return -EINVAL;
+	}
+
+	if (config->beam_off_defined > 1) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid beam_off_defined %d",
+			    config->beam_off_defined);
+		return -EINVAL;
+	}
+
+	if (config->num_mic_locations > SOF_TDFB_MAX_MICROPHONES) {
+		comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid num_mic_locations %d",
+			    config->num_mic_locations);
+		return -EINVAL;
+	}
+
+	if (config->num_angles == 0 && config->num_mic_locations == 0) {
+		legacy = 1;
+		num_filters = config->num_filters;
+	} else {
+		num_filters = config->num_filters * (config->num_angles + config->beam_off_defined);
+	}
+
+	/* Skip filter coefficients */
+	coefp = tdfb_filter_seek(config, num_filters);
+
+	/* Get shortcuts to input and output configuration */
+	cd->input_channel_select = coefp;
+	coefp += config->num_filters;
+	cd->output_channel_mix = coefp;
+	coefp += config->num_filters;
+	cd->output_stream_mix = coefp;
+	coefp += config->num_filters;
+
+	if (legacy) {
+		cd->filter_angles = NULL;
+		cd->mic_locations = NULL;
+		if ((char *)coefp != (char *)config + config->size) {
+			comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid config size");
+			return -EINVAL;
+		}
+	} else {
+		if (config->beam_off_defined) {
+			output_channel_mix_beam_off = coefp;
+			coefp += config->num_filters;
+		}
+		cd->filter_angles = (struct sof_tdfb_angle *)coefp;
+		cd->mic_locations = (struct sof_tdfb_mic_location *)
+			(&cd->filter_angles[config->num_angles]);
+		if ((char *)&cd->mic_locations[config->num_mic_locations] !=
+		    (char *)config + config->size) {
+			comp_cl_err(&comp_tdfb, "tdfb_init_coef(), invalid config size");
+			return -EINVAL;
+		}
+	}
+
+	/* Skip to requested coefficient set */
+	if (legacy) {
+		idx = 0;
+		comp_cl_info(&comp_tdfb, "tdfb_init_coef(), configure legacy mode");
+	} else {
+		min_delta = 360;
```

Comment:
btw, should we be using radians here ?. (given that POSIX trig APIs all use radians) ?

Reply:
It would be then e.g. Q22.10 radians. In this case I thought degrees (as Q32.0 integers) would be easier for humans but the implementation could be both. The direction tracker algorithm will be radians internally, at least.

How would  you prever the enums, it's now in text:

  ```
  ; Item #0 '0deg'
  ; Item #1 '30deg'
  ; Item #2 '60deg'
  ; Item #3 '90deg'
  ; Item #4 '120deg'
  ; Item #5 '150deg'
  ; Item #6 '180deg'
  ; Item #7 '210deg'
  ; Item #8 '240deg'
  ; Item #9 '270deg'
  ; Item #10 '300deg'
  ; Item #11 '330deg'

```

In radians the controls would be fractions of pi, e.g.  30 degrees "1/6*pi", "2/6*pi". It's harder for me at least to figure out what it is visually as beam angle. Degrees in enums and radians in internal database could be suitable compromise. What do you think?

Reply:
Thanks got you .Btw, do we need more granularity here than 30deg ? i.e. would a volume control be better with 0 - 359 ?

Reply:
The configuration blob contains the multiplier and angle offset from enum value to degrees. Currently alsatplg limits the number of steps to 16. According to @juimonen there does not seem to be limitation in kernel so maybe even 360 enums could be possible. But it would need ALSA change.

Yep, volume control could be possible. In topology it should bypass the decibels definition and set up somehow to show meaningful values in alsamixer. I don't know if that should be tried or not. With low amount of microphones (8 or less) the beams are wider than the current granularity 30 degrees.

But once I'm past [WIP] with this with remaining issues I'll try how it would look like.